runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,298 @@
+
+        /**
+         * It is ok to check equality of a CorfuRecordId with ordering data against one without.
+         * @param o object to compare against.
+         * @return
+         */
+        public boolean equals(CorfuRecordId o) {
+            return id.getLeastSignificantBits() == o.id.getLeastSignificantBits();
+        }
+
+        public String toString() {
+            return id.toString();
+        }
+    }
+
+    /**
+     * Appends the specified element at the end of this unbounded queue.
+     * In a distributed system, the order of insertions cannot be guaranteed
+     * unless a transaction is used.
+     * Capacity restrictions and backoffs must be implemented outside this
+     * interface. Consider validating the size of the queue against a high
+     * watermark before enqueue.
+     *
+     * @param e the element to add
+     * @throws IllegalArgumentException if some property of the specified
+     *         element prevents it from being added to this queue
+     * @return Unique ID representing this entry in the persistent queue
+     * WARNING: The ID returned by this function will NOT be the comparable with the ID
+     *          returned by the CorfuQueueRecord from entryList() method because if this
+     *          method executes within a transaction, the comparable ordering is not
+     *          known until the transaction commits.
+     *          The ID returned here is only really useful for remove() operations.
+     */
+    public CorfuRecordId enqueue(E e) {
+        final Long id = guidGenerator.nextLong();
+        corfuTable.put(id, e);
+        return new CorfuRecordId(0, id);
+    }
+
+    /**
+     * CorfuQueueRecord encapsulates each entry enqueued into CorfuQueue with its unique ID.
+     * It is a read-only type returned by the entryList() method.
+     * The ID returned here can be used for both point get()s as well as remove() operations
+     * on this Queue.
+     *
+     * @param <E>
+     */
+    public class CorfuQueueRecord<E> implements Comparable<CorfuQueueRecord<E>> {
+        /**
+         * This ID represents the entry and its order in the Queue.
+         * This implies that it is unique and comparable with other IDs
+         * returned from CorfuQueue methods with respect to its enqueue order.
+         * However it cannot be compared with ID returned from the enqueue method for ordering
+         * because if this method is wrapped in a transaction, the order is established only later.
+         */
+        @Getter
+        final private CorfuRecordId recordId;
+
+        @Getter
+        final private E entry;
+
+        public String toString() {
+            return String.format("%s=>%s", recordId, entry);
+        }
+
+        CorfuQueueRecord(long ordering, long entryId, E entry) {
+            this.recordId = new CorfuRecordId(ordering, entryId);
+            this.entry = entry;
+        }
+
+        @Override
+        public int compareTo(CorfuQueueRecord<E> o) {
```

Comment:
Is it better to have the parameter of compareTo being CorfuQueueRecord<? extends E>?

Reply:
this is a great suggestion! Thank you!